# Ederson
# Identification of primer targets for identifying biomarkers for specific bacteria

:microscope: :thumbsup:

###  By Jared Flater in collaboration with Ederson Jesus and Adina Howe

# Inputs: List of accession numbers of targets in Accession_numbers_for_Jared_and_Adina.xlsx
We are interested in IDing unique sequences in R.tropici and R.freirei that are not found in other rhizobium species as well as not found in RefSoil db or RefSeq db. Once we have those identified, we will id 19 bp seq to target for primer design. 
![](Images/ideas.jpg)

----

# G1:Rhizobium tropici CIAT 899 (+chromosome) 
1. NC_020059.1+
2. NC_020060.1
3. NC_020061.1
4. NC_020062.1

----

# G2:Rhizobium freirei PRF 81
5. NZ_AQHN01000095.1
6. NZ_AQHN0100009 6.1
7. NZ_AQHN1000084.1 

----
#  Download RefSoilpython k-mer.py from https://figshare.com/articles/RefSoil_Database

#  First script in k-mer proce1ss:

```python

"""breaks up a sequence from a .fa file into kmerlength, and compares to kmers from other genomes"""
#python k-mer.py [reference1.fa] [comparison1.fa] [comparison.fa] [comparison3.fa] > filename.fa

import sys
import screed

def consume_genome(fname):
    """reads through a .fa file, returning names and sequences"""
    genome = ''
    for i in fname:
        for record in screed.open(i):
            genome = genome + record.sequence
    return genome

def kmer_count(genomelookingin, kmerlength):
    """makes a dictionary of 150 bp kmers"""
    kmer150 = {}
    for i in range(len(genomelookingin)+1-kmerlength):
        kmer = genomelookingin[i:i+kmerlength]
        if kmer150.has_key(kmer):
            continue
        else:
            kmer150[kmer] = kmer150.get(kmer, 0)+1
    return kmer150

def rolling_window(seq, window_size):
    """rolling window slides along a sequence, showing only 150 bp at a time"""
    for i in xrange(len(seq) - window_size + 1):
        yield seq[i:i+window_size]


def main():
    """main for k-mer.py, find kmers in a .fa file."""
    fname = sys.argv[1:2]
    fname_compare = sys.argv[2:]


    ref_genome1 = consume_genome(fname)
    kmer_dict = kmer_count(ref_genome1, 150)
    ref_genome2 = consume_genome(fname_compare)

    for seq in rolling_window(ref_genome2, 150):
        if kmer_dict.has_key(seq):
            del kmer_dict[seq]

    for i, kmer in enumerate(kmer_dict.keys()):
        print ">" + str(i) + "_" + fname[0]
        print kmer

if __name__ == '__main__':
    main()
```

```
bash-3.2$ python Scripts/k_mer.py test/test_NC.fa test/test_NZ.fa test/test_compare.fa > 1_NC_kmer_test.fa
```
# For each file output looks like:
```
bash-3.2$ head 1_NC_kmer_test.fa 
>0_test/test_NC.fa
ACTATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGC
>1_test/test_NC.fa
CCATCAGAACACGCGGCGTGATCTGCCAGGATACGCCCCATCTGTCCTTGCACCAGCCGCATTCGCTTTCCTGGCCGCCATTGCCGACAATGGCGTTCCAATAACGATCCGTCTCTTCCTGGTCCTCGGTGTAGATCTGAAACGAAAAGG
>2_test/test_NC.fa
ATACCAATAAGACTTTGAAAAAGCTGGCGGCCCGAGGGTTGCTCCTTTGGCAGGAGAGGGGCTGCGAGGTTCTGGATGGGAAGGGGCTTGCAGAACTTGCGGGCTGGGACGGTTTTGATTCGGGCAAGCGCCCCTTCATCTAGGCACTGC
>3_test/test_NC.fa
AAGGCAGGTAAAACGGGTGAAAGATTCCTGTCGATACTCATTGTCCTGCATAAAATCGACATTGCTGTCGTAGATCTCATATTTGAGGCGGTCTTCCACATATTCTTGCGGTCGAGATCGCTGCTGCTTTTGCATAGCAGTGAGCGGCCG
>4_test/test_NC.fa
TGCCGCATCGCCTCGATCGCCATGCAGCAGATGTCATTGCCGCGGCCAATGGTGAAGGTCAGGCCGTGGCCAGCAAGGCCCGGCTTGTCGGTGTCGAGGATGACATAAGCTGCCGAATAATCCGGATCCGGATTCATCGCATCCGAACCA
```
This generated a file that was over 1 gb in size, I wasn't really expecting a file this large, however, it may be possible. To test this, I made three files... one .fa file that contained the first 150bp kmer + 5bp of random. The second file is just a .fa with zzzzzzzzzzzzz filling it...the final is a .fa with the same first 150 bp as the first file. 

When we compare 155pb_test.fa to 155zbp_test.fa and 150bp_test.fa, we should get 5 kmers:

```
jflater-air:test jaredflater$ python ../Scripts/k_mer.py 155pb_test.fa 155zbp_test.fa 150bp_test.fa
>0_155pb_test.fa
ATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGT
>1_155pb_test.fa
TATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGTT
>2_155pb_test.fa
CTATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCG
>3_155pb_test.fa
TATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGG
>4_155pb_test.fa
ATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGTTC
```
Now, if we make the same comparison, less the final 150bp_test.fa file, we should have 6 kmers returned:

```
jflater-air:test jaredflater$ python ../Scripts/k_mer.py 155pb_test.fa 155zbp_test.fa
>0_155pb_test.fa
ACTATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGC
>1_155pb_test.fa
ATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGT
>2_155pb_test.fa
TATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGTT
>3_155pb_test.fa
CTATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCG
>4_155pb_test.fa
TATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGG
>5_155pb_test.fa
ATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGCGGTTC
```

# Now we will move onto the second script in the process, identifying potential primers for each kmer, ensuring that each primer is only present in one gene.  This scripts input is the k-mers unique to each genome and output is the potential primers at each end of kmer:

```python
"""Takes in kmers and removes the front and back 19 bp"""
#python pri-mer.py

import sys
import screed

def rolling_window(seq, window_size):
    """rolling window slides along a sequence, in this case showing 19 bp at a time"""
    for i in xrange(len(seq) - window_size + 1):
        yield seq[i:i+window_size]

def main():
    """main for k-mer.py, find kmers in a .fa file."""
    #reading in your kmer file
    for record in screed.open(sys.argv[1]):
        my_seq = record.sequence
    #print my_seq
        for i, seq in enumerate(rolling_window(my_seq, 19)):
            if i == 0:
                print ">" + str(i) + "_" + record.name + "_forward"
                print seq
            if i == 131:
                print ">" + str(i) + "_" + record.name + "_reverse"
                print seq

if __name__ == '__main__':
    main()
```
# Run it like this: 
```
jflater-air:test jaredflater$ python ../Scripts/pri_mer.py 150_kmer_test.fa > 150_primer_test.fa
```
# This is the output from pri_mer.py on 150_kmer_test.fa
```
jflater-air:test jaredflater$ cat 150_primer_test.fa
>0_0_155pb_test.fa_forward
ACTATATCGTGGTTTGCAC
>131_0_155pb_test.fa_reverse
GTGCTTAACATTGCCCGGC
>0_1_155pb_test.fa_forward
ATATCGTGGTTTGCACTGT
>131_1_155pb_test.fa_reverse
CTTAACATTGCCCGGCGGT
>0_2_155pb_test.fa_forward
TATCGTGGTTTGCACTGTG
>131_2_155pb_test.fa_reverse
TTAACATTGCCCGGCGGTT
>0_3_155pb_test.fa_forward
CTATATCGTGGTTTGCACT
>131_3_155pb_test.fa_reverse
TGCTTAACATTGCCCGGCG
>0_4_155pb_test.fa_forward
TATATCGTGGTTTGCACTG
>131_4_155pb_test.fa_reverse
GCTTAACATTGCCCGGCGG
>0_5_155pb_test.fa_forward
ATCGTGGTTTGCACTGTGA
>131_5_155pb_test.fa_reverse
TAACATTGCCCGGCGGTTC
```
# Now we need to get the "body" out of each k-mer (the middle is 112 bp, 150 total bp minus the two 19 bp ends)
# For this, use body.py

```python
"""use this to feed the body (150-(19*2))bp into a body.fa. We need to make sure no primers match any portion of the k-mer. In addtion, we also include this with all other .fa files we are comparing to"""
#python body.py

import sys
import screed

def rolling_window(seq, window_size):
    """Breaks down a sequence input into corresponding k-mers"""
    for i in xrange(len(seq) - window_size + 1):
        yield seq[i:i+window_size]

def main():
    """main for body.py"""
    for record in screed.open(sys.argv[1]):
        my_seq = record.sequence
        for i, seq in enumerate(rolling_window(my_seq, 112)):
            if i == 19:
                print ">" + str(i) + "_" + record.name
                print seq

if __name__ == '__main__':
    main()
```
Run body.py:
```
jflater-air:test jaredflater$ python ../Scripts/body.py 150_kmer_test.py > body_test.fa
```
# This is what the output of body.py looks like:
```
jflater-air:test jaredflater$ cat body_test.fa
>19_0_155pb_test.fa
TGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAG
>19_1_155pb_test.fa
GACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTG
>19_2_155pb_test.fa
ACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGC
>19_3_155pb_test.fa
GTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGG
>19_4_155pb_test.fa
TGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGT
>19_5_155pb_test.fa
CGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCT
```
# if we paste the k-mer, primer, and body into text edit, we can see that everything lines up how would like it to :
```
>0_155pb_test.fa
ACTATATCGTGGTTTGCACTGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAGGTGCTTAACATTGCCCGGC
>0_0_155pb_test.fa_forward                                                                                                >131_0_155pb_test.fa_reverse
ACTATATCGTGGTTTGCAC                                                                                                                GTGCTTAACATTGCCCGGC
>19_0_155pb_test.fa
                   TGTGACGGGTTCGACGCAAGCCGGCATGGTCGTTGGTTTCGCCAAGGATGGACGACAGCGCAATGTGATCGGTATCGATGCTTCGGCAACCCCTCTCCAAGCCCAGTCGCAG
```
## Work on validating the primers:
```
bash-3.2$ python ../Scripts/primer_validate.py 150_primer_test.py test_compare.fa > validated_primers.fa
```
```python
"""similar to k_mer.py, but for primers from those kmers"""
#python primer_validate.py [primer.fa] [comparison1.fa] [comparison.fa] [comparison3.fa] > name.fa

#To take in primers forward and back
#And validate them by making sure only appear in genome of interest

import sys
import screed

#Not sure on this script, needs testing, was formated from k_mer.py
def consume_genome(fname):
    """reads through a .fa file, returning names and sequences"""
    genome = ''
    for i in fname:
        for record in screed.open(i):
            genome = genome + record.sequence
    return genome

def make_kmer_dict(fname):
    """making an object with names and sequences of a .fa"""
    for i in fname:
        for record in screed.open(i):
            print record.name
            print record.sequence

def kmer_count(kmerlookingin, primerlength):
    """makes dictionary of primers"""
    primer19 = {}
    print kmerlookingin
    for i in range(len(kmerlookingin)+1-primerlength):
        kmer = kmerlookingin[i:i+primerlength]
        if primer19.has_key(kmer):
            continue
        else:
            primer19[kmer] = 1
    return primer19

#breaks a sequence into its respective k-mers
def rolling_window(seq, window_size):
    """rolling window slides along a sequence, showing 19 bp at a time"""
    for i in xrange(len(seq) - window_size + 1):
        yield seq[i:i+window_size]

def make_primer_dict(fname):
    """primer dict function"""
    i = {}
    for i in fname:
        for record in screed.open(i):
            i[record.sequence] = record.name
    return i

def main():
    """main for primer validation"""
    fname = sys.argv[1:2]
    fname_compare = sys.argv[2:]

    kmer_dict = make_primer_dict(fname)
    #kmer_dict is a dictionary of all 19-mers in fnames
    ref_genome2 = consume_genome(fname_compare)
    #breakdown fname_compare into 19-mers
    for seq in rolling_window(ref_genome2, 19):
        if kmer_dict.has_key(seq):
            del kmer_dict[seq]

    for i, kmer in enumerate(kmer_dict.keys()):
        print ">" + str(i) + "_" + kmer_dict[kmer]
        print kmer

if __name__ == '__main__':
    main()
```
# 6/15/2017 output:
```
bash-3.2$ tail validated_primers.faclear 
>203_150_primer_test.py
TCTTAACATTGCCCGGCGG
>204_150_primer_test.py
ATATCGTGGTTTGCACTGT
>205_150_primer_test.py
GCACTGGCTTAACATTGCC
>206_150_primer_test.py
ATTGCCCGGCGGTTATCGT
>207_150_primer_test.py
CTTGCTTAACATTGCCCGG
bash-3.2$ head validated_primers.faclear                                                 
0_0_155pb_test.fa_forward
ACTATATCGTGGTTTGCAC
131_0_155pb_test.fa_reverse
GTGCTTAACATTGCCCGGC
0_1_155pb_test.fa_forward
ATATCGTGGTTTGCACTGT
131_1_155pb_test.fa_reverse
CTTAACATTGCCCGGCGGT
0_2_155pb_test.fa_forward
TATCGTGGTTTGCACTGTG
```
# Work on complement for each primer:

```python
"""Find reverse complement of reverse primer, print forward primer"""
# Input file should be in .fa format, with both forward and reverse primer
# the word "reverse" in record.name of each reverse primer
# Make sure we do the reverse complement of the forward primers too
import sys
import screed

from Bio.Seq import Seq

def main():
    """this is the main, see above"""
    for record in screed.open(sys.argv[1]):
        seq = record.sequence
        my_seq = Seq(seq)
        if "reverse" in record.name:
            print ">" + record.name + "_complement"
            print my_seq.reverse_complement()
        else:
            print ">" + record.name
            print my_seq

if __name__ == '__main__':
    main()
```
##  Run subset fasta.py on all files in "full genomes" and "master genomes"
```
for file in compare_genomes/*.fa; do python subset_fasta.py $file >> Ederson-hpc/test/test_compare_genomes.fa; done
```
make sure file is empty or does not exist when running above ^
# Handy script to return odd lines in a file:
```
sed -n 1~2p file
```
This will start at line 1 and print every 2 lines, odd ones. If this were a fasta file, it would only print the carrot lines!
```
[flaterj1@dev-intel14 test]$ sed -n 1~2p test_compare.fa 
[flaterj1@dev-intel14 test]$ head test_compare.fa
>NC_004041.2 Rhizobium etli CFN 42 plasmid symbiotic plasmid p42d, complete sequence
>NC_007761.1 Rhizobium etli CFN 42, complete genome
>NC_007762.1 Rhizobium etli CFN 42 plasmid p42a, complete sequence
>NC_007763.1 Rhizobium etli CFN 42 plasmid p42b, complete sequence
>NC_007764.1 Rhizobium etli CFN 42 plasmid p42c, complete sequence
>NC_007765.1 Rhizobium etli CFN 42 plasmid p42e, complete sequence
>NC_007766.1 Rhizobium etli CFN 42 plasmid p42f, complete sequence
>NC_008378.1 Rhizobium leguminosarum bv. viciae plasmid pRL12 complete genome, strain 3841
>NC_008379.1 Rhizobium leguminosarum bv. viciae plasmid pRL9 complete genome, strain 3841
>NC_008380.1 Rhizobium leguminosarum bv. viciae chromosome complete genome, strain 3841
```

# Need to investigate how to test a primer for it's reverse compliment

```
cat input.fa | while read L; do echo $L; read L; echo "$L" | rev | tr "ATGC" "TACG" ; done
```
# I seems that for this to work on a file with multiple lines, the .fa file should first be linearized:
```
awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' < input.fa
```
## To remove large files stuck in the commit history of git:
```
git filter-branch --index-filter 'git rm -r --cached --ignore-unmatch <file/dir>' HEAD
```
# Run a qsub:
```
$ qsub mysub.qsub 
```
# Example shell commands and file names for first script:
----
Find unique k-mers in r.freire
```
python k-mer.py refrence_genomes/NZ_AQHN01000095.1.fa refrence_genomes/NZ_AQHN01000096.1.fa refrence_genomes/NC*.fa compare_genomes/*.fa > r.freirei.mers.fa
```
Find unique k-mers in r.tropici
```
phython k-mer.py refrence_genomes/NC_02006*.fa refrence_genomes/NC_020059.1.fa refrence_genomes/NZ*.fa compare_genomes/*.fa > r.tropici.mers.fa
```
